### PR TITLE
Adds context parallelism utilities: moving cp shards to diff ranks and pad sequence to divisibility factory

### PR DESCRIPTION
**Pull request by jomitchellnv**
# Description

In order to utilize the context parallelism functionality as part of TransformerEngine we must perform some manipulation on the training data and its metadata to make it ammendable towards context parallelism.

In doing so, we must pad our sequences to be divisible by `cp_size*2` so that we are able to perform the dual chunking method. This utility function is now added in this MR, such that it will pad each sequence to make it divisible.

Moreover, we also add the functionality to move data to each CP rank. For example if you have a sequence
```
Sequence=[1, 2, 3, 4, 5, 6, 7, 8]
```
The function `get_thd_batch_on_this_cp_rank` will move data to each shard
```
cp0: [1, 2, 7, 8]
cp1: [3, 4, 5, 6]
```
So that its easy to use ring attention on it. 

Here I've added both utility functions as well as rigorous unit testing to ensure these functions are durable.

Fixes # (issue)

## Type of change

- [ ] Documentation change (change only to the documentation, either a fix or a new content)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Infra/Build change
- [ ] Code refactoring

## Changes

Please list the changes introduced in this PR:

- Adds the function get_thd_batch_on_this_cp_rank for CP sharding
- Adds the function pad_sequences_to_divisibility to augment training data to make it amenable toward context parallelism.

# Checklist:

- [X] I have read and followed the [contributing guidelines](https://github.com/NVIDIA/TransformerEngine/blob/main/CONTRIBUTING.rst)
- [X] The functionality is complete
- [X] I have commented my code, particularly in hard-to-understand areas
- [no] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [X] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: transformer_engine/pytorch/attention/dot_product_attention/context_parallel.py [hunk 1]
```diff
@@ -48,6 +48,266 @@
...
+        - cu_seqlens_q_padded: Cumulative sequence lengths accounting for padding
+    """
+    # Flatten input_ids and labels if needed
+    if input_ids.dim() == 2:
+        input_ids = input_ids.squeeze(0)
+    if labels.dim() == 2:
+        labels = labels.squeeze(0)
+
+    # Extract individual sequences
+    batch_sequences = extract_sequences(input_ids, cu_seqlens_q)
```
Comment (sudhakarsingh27) [inline transformer_engine/pytorch/attention/dot_product_attention/context_parallel.py]: These and other funcs are defined later. They should be either defined earlier or if their usage scope is purely `pad_sequences_to_divisibility`, then perhaps they should be inner functions?

File: transformer_engine/pytorch/attention/dot_product_attention/context_parallel.py [hunk 2]
```diff
@@ -48,6 +48,266 @@
 _seq_chunk_ids_cache_for_reordering_after_attn = {}
 
 
+def pad_sequences_to_divisibility(
+    input_ids: torch.Tensor,
+    labels: torch.Tensor,
+    positional_ids: torch.Tensor,
+    cu_seqlens_q: torch.Tensor,
```
Comment (sudhakarsingh27) [inline transformer_engine/pytorch/attention/dot_product_attention/context_parallel.py]: nitpick: `cu_seqlens_q` -> should it be `cu_seqlens` instead?

File: transformer_engine/pytorch/attention/dot_product_attention/context_parallel.py [hunk 3]
```diff
@@ -48,6 +48,266 @@
 _seq_chunk_ids_cache_for_reordering_after_attn = {}
 
 
+def pad_sequences_to_divisibility(
```
Comment (sudhakarsingh27) [inline transformer_engine/pytorch/attention/dot_product_attention/context_parallel.py]: nitpick: maybe a better name is `pad_thd_sequences_for_cp` (or `pad_packed_sequences_for_cp`) for more clarity?

File: transformer_engine/pytorch/attention/dot_product_attention/context_parallel.py [hunk 4]
```diff
@@ -48,6 +48,266 @@
...
+            " each sequence. cu_seqlens_q_padded is required for THD format to calculate the right"
+            " chunk sizes for each sequence."
+        )
+
+        # Get the cu_seqlens_kv_padded
+        cu_seqlens_kv_padded = cu_seqlens_kv_padded
+        assert cu_seqlens_kv_padded is not None, (
+            "cu_seqlens_kv_padded is required for THD format to calculate the right chunk sizes for"
+            " each sequence."
+        )
```
Comment (sudhakarsingh27) [inline transformer_engine/pytorch/attention/dot_product_attention/context_parallel.py]: Unused variable

File: transformer_engine/pytorch/attention/dot_product_attention/context_parallel.py [hunk 5]
```diff
@@ -48,6 +48,266 @@
...
+    padded_labels = pad_batch_sequences(batch_labels, padding_amounts, padding_label_id)
+
+    # Create padded input_ids tensor
+    input_ids_padded = torch.cat(padded_sequences)
+
+    # Create padded labels tensor
+    labels_padded = torch.cat(padded_labels)
+
+    # Create padded positional_ids (positions continue incrementing through padding)
+    positional_ids_padded = create_padded_positional_ids(
```
Comment (sudhakarsingh27) [inline transformer_engine/pytorch/attention/dot_product_attention/context_parallel.py]: Having `positional_ids` here seems redundant since `cu_seqlens...` could be used to derive it. We could avoid the input `positional_ids` and return newly derived `positional_ids` tensor (it could be computed in a simpler way)

File: transformer_engine/pytorch/attention/dot_product_attention/context_parallel.py [hunk 6]
```diff
@@ -48,6 +48,266 @@
...
+    cu_seqlens_q: torch.Tensor,
+    divisibility_factor: int,
+    padding_token_id: int = 0,
+    padding_label_id: int = -100,
+) -> Tuple[torch.Tensor, torch.Tensor, torch.Tensor, torch.Tensor, torch.Tensor]:
+    """Pads sequences to be divisible by the divisibility factor.
+
+    Args:
+        input_ids: Tensor of shape (1, N) containing concatenated sequences
+        labels: Tensor of shape (1, N) containing labels for each token
```
Comment (sudhakarsingh27) [inline transformer_engine/pytorch/attention/dot_product_attention/context_parallel.py]: Shapes for `input_ids` and `labels` in this function are assumed to be (N,) and not `(1, N)`

**Review by sudhakarsingh27 — CHANGES_REQUESTED** [review]
Thanks for adding these APIs are corresponding tests. A few changes that I feel would be helpful:
- Please append the new APIs at the end of the file `context_parallel.py`
- While we're at it, is it also possible to make `pad_sequences_to_divisibility` function a little less verbose like following (could also help avoid creating a lot of helper functions)?

```
# Compute the sequence lengths from cu_seqlens
seqlens = cu_seqlens[1:] - cu_seqlens[:-1]

# Function to compute required padding for each sequence (make length a multiple of 2*CP_SIZE)
def required_padding(seqlen):
    chunk = 2 * CP_SIZE
    padded_length = ((seqlen // chunk) + 1) * chunk
    return padded_length - seqlen

# List: amount of padding needed for each sequence
padding_amounts = [required_padding(l) for l in seqlens]

# Extract sequences and labels for each batch item
batch_sequences = [input_ids[start:end] for start, end in zip(cu_seqlens[:-1], cu_seqlens[1:])]
batch_labels = [labels[start:end] for start, end in zip(cu_seqlens[:-1], cu_seqlens[1:])]

# Pad sequences and labels to required length
def pad_tensor(seq, pad_amount):
    return torch.cat([seq, torch.zeros(pad_amount)])

padded_sequences = torch.cat([pad_tensor(seq, pad) for seq, pad in zip(batch_sequences, padding_amounts)])
padded_labels = torch.cat([pad_tensor(seq, pad) for seq, pad in zip(batch_labels, padding_amounts)])

# Compute cumulative padded sequence lengths, starting from 0
padded_lengths = seqlens + torch.tensor(padding_amounts)
padded_cu_seqlens = torch.cumsum(
    torch.cat([torch.tensor([0]), padded_lengths]), dim=0
)
print(padded_cu_seqlens)

# Compute revised positional IDs for all batch sequences
padded_positional_ids = torch.cat([
    torch.arange(0, int(length)) for length in padded_lengths
])
padded_positional_ids
```

File: transformer_engine/pytorch/attention/dot_product_attention/context_parallel.py (hunk 1 above)
Comment (jomitchellnv) [reply transformer_engine/pytorch/attention/dot_product_attention/context_parallel.py]: I removed this function.

File: transformer_engine/pytorch/attention/dot_product_attention/context_parallel.py (hunk 2 above)
Comment (jomitchellnv) [reply transformer_engine/pytorch/attention/dot_product_attention/context_parallel.py]: fixed

File: transformer_engine/pytorch/attention/dot_product_attention/context_parallel.py (hunk 3 above)
Comment (jomitchellnv) [reply transformer_engine/pytorch/attention/dot_product_attention/context_parallel.py]: sure done.

File: transformer_engine/pytorch/attention/dot_product_attention/context_parallel.py (hunk 4 above)
Comment (jomitchellnv) [reply transformer_engine/pytorch/attention/dot_product_attention/context_parallel.py]: yea nice catch. I removed it

File: transformer_engine/pytorch/attention/dot_product_attention/context_parallel.py (hunk 6 above)
Comment (jomitchellnv) [reply transformer_engine/pytorch/attention/dot_product_attention/context_parallel.py]: I'll update this so that they can be either which is why I squeeze them below which may make this function more robust.

File: transformer_engine/pytorch/attention/dot_product_attention/context_parallel.py (hunk 5 above)
Comment (jomitchellnv) [reply transformer_engine/pytorch/attention/dot_product_attention/context_parallel.py]: Yea you're totally right!! I removed the positional_id padding part of this function and then created a separate function for it, which I then added unit tests for.

File: transformer_engine/pytorch/attention/dot_product_attention/context_parallel.py [hunk 7]
```diff
@@ -3927,3 +3927,215 @@ def attn_forward_func_with_cp(
...
+    Which are parallelized across GPUs in a context parallel group.
+    This version works with variable-length sequences using cumulative sequence lengths.
+    """
+    # Get context parallel size and rank
+    cp_size = torch.distributed.get_world_size(group=cp_group)
+    if cp_size > 1:
+        cp_rank = torch.distributed.get_rank(group=cp_group)
+
+        # Get cumulative sequence lengths from batch
+        cu_seqlens_padded = cu_seqlens_q_padded
```
Comment (sudhakarsingh27) [inline transformer_engine/pytorch/attention/dot_product_attention/context_parallel.py]: should we change the argument name to this method then, if it's not used anywhere

Comment (jomitchellnv) [reply transformer_engine/pytorch/attention/dot_product_attention/context_parallel.py]: fixed

File: transformer_engine/pytorch/attention/dot_product_attention/context_parallel.py [hunk 8]
```diff
@@ -3927,3 +3927,207 @@ def attn_forward_func_with_cp(
...
+        total_slices_of_any_sequence = 2 * cp_size
+        slice_sizes = (
+            cu_seqlens_padded[1:] - cu_seqlens_padded[:-1]
+        ) // total_slices_of_any_sequence
+
+        # Process each tensor directly instead of using keys_to_change loop
+        def process_tensor(val):
+            if val is not None:
+                # Determine which dimension is the sequence dimension
+                # TODO: Add support for if the cu_seqlens_padded gives you a tensor of shape (1,)
```
Comment (sudhakarsingh27) [inline transformer_engine/pytorch/attention/dot_product_attention/context_parallel.py]: Pylint won't let this pass. Add an assertion instead

File: transformer_engine/pytorch/attention/dot_product_attention/context_parallel.py [hunk 9]
```diff
@@ -3927,3 +3927,207 @@ def attn_forward_func_with_cp(
...
+    # Generate positional IDs for each padded sequence (each starts from 0)
+    padded_lengths = seqlens + torch.tensor(padding_amounts, dtype=seqlens.dtype)
+    positional_ids = torch.cat(
+        [torch.arange(0, int(length), dtype=dtype) for length in padded_lengths]
+    )
+
+    return positional_ids
+
+
+def get_thd_batch_on_this_cp_rank(
```
Comment (sudhakarsingh27) [inline transformer_engine/pytorch/attention/dot_product_attention/context_parallel.py]: Based on our discussion offline, we should rename this to `get_data_batch_on_this_cp_rank`(or similar) and pass  `qkv_format=[thd,bshd,sbhd]` as an argument to differentiate the cases. Since we haven't tested non-THD, we could add an error assertion for the case when `qkv_format=bshd/sbhd` for now and later add support for it with test cases.

Comment (jomitchellnv) [reply transformer_engine/pytorch/attention/dot_product_attention/context_parallel.py]: ok. I see

File: transformer_engine/pytorch/attention/dot_product_attention/context_parallel.py [hunk 10]
```diff
@@ -3927,3 +3927,212 @@ def attn_forward_func_with_cp(
...
+                    )
+
+                return val.index_select(current_seq_dim, torch.cat(cp_rank_slices))
+
+            # Process each tensor directly
+            input_ids_padded = process_tensor(input_ids_padded)
+            labels_padded = process_tensor(labels_padded)
+            position_ids_padded = process_tensor(position_ids_padded)
+    else:
+        raise ValueError(f"Unsupported qvk_format: {qvk_format}!")
```
Comment (sudhakarsingh27) [inline transformer_engine/pytorch/attention/dot_product_attention/context_parallel.py]:
```suggestion
        raise ValueError(f"Support not implemented yet for qvk_format: {qvk_format}!")
```

**Review by sudhakarsingh27 — APPROVED** [review]
Apart from the nitpicks, looks good. I'd wait for the CI to finish before merging

File: transformer_engine/pytorch/attention/dot_product_attention/context_parallel.py (hunk 10 above)
Comment (jomitchellnv) [reply transformer_engine/pytorch/attention/dot_product_attention/context_parallel.py]: updated
